sdk/core/scheduler/timer.h

```diff
@@ -26,36 +26,96 @@ namespace
 	  IsTimer<TimerCore>,
 	  "Platform's timer implementation does not meet the required interface");
 
+	/**
+	 * Timer interface.  Provides generic timer functionality to the scheduler,
+	 * wrapping the platform's timer device.
+	 */
 	class Timer final : private TimerCore
 	{
+		inline static uint64_t lastTickTime         = 0;
+		inline static uint64_t zeroTickTime         = 0;
+		inline static uint32_t accumulatedTickError = 0;
+
 		public:
+		/**
+		 * Perform any setup necessary for the timer device.
+		 */
 		static void interrupt_setup()
 		{
 			static_assert(TIMERCYCLES_PER_TICK <= UINT32_MAX,
 			              "Cycles per tick can't be represented in 32 bits. "
 			              "Double check your platform config");
 			init();
-			setnext(TIMERCYCLES_PER_TICK);
+			zeroTickTime = time();
 		}
 
-		static void do_interrupt()
-		{
-			++Thread::ticksSinceBoot;
+		/**
+		 * Expose the timer device's method for returning the current time.
+		 */
+		using TimerCore::time;
 
-			expiretimers();
-			setnext(TIMERCYCLES_PER_TICK);
+		/**
+		 * Update the timer to fire the next timeout for the thread at the
+		 * front of the queue, or disable the timer if there are no threads
+		 * blocked with a timeout and no threads with the same priority.
+		 *
+		 * The scheduler is a simple RTOS scheduler that does not allow any
+		 * thread to run if a higher-priority thread is runnable.  This means
+		 * that we need a timer interrupt in one of two situations:
+		 *
+		 *  - We have a thread of the same priority as the current thread and
+		 *    we are going to round-robin schedule it.
+		 *  - We have a thread of a higher priority than the current thread
+		 *    that is currently sleeping on a timeout and need it to preempt the
+		 *    current thread when its timeout expires.
+		 *
+		 * We currently over approximate the second condition by making the
+		 * timer fire independent of the priority.  If this is every changed,
+		 * some care must be taken to ensure that dynamic priority propagation
+		 * via priority-inheriting futexes behaves correctly.
+		 *
+		 * This should be called after scheduling has changed the list of
+		 * waiting threads.
+		 */
+		static void update()
+		{
+			auto *thread             = Thread::current_get();
+			bool  waitingListIsEmpty = ((Thread::waitingList == nullptr) ||
+                                       (Thread::waitingList->expiryTime == -1));
+			bool  threadHasNoPeers =
+			  (thread == nullptr) || (!thread->has_priority_peers());
+			if (waitingListIsEmpty && threadHasNoPeers)
+			{
+				clear();
+			}
+			else
+			{
+				uint64_t nextTimer = waitingListIsEmpty
```

Comment:
@davidchisnall if the waiting list is non-empty *and* the current thread has peers I think this will set the timer for the next expiry time. Doesn't that mean the current thread could get more than one quantum?

Reply:
Fixed in #240, thanks!